Import HCA metadata into Neo4j, MongoDB, Elasticsearch

mruffalo opened the pull request:
This PR adds a `manage.py` command to load a metadata spreadsheet from HCA into Neo4j, MongoDB, and Elasticsearch. (Such a spreadsheet can be obtained [from pages like this](https://data.humancellatlas.org/explore/projects/8c3c290d-dfff-4553-8868-54ce45f4ba7f)).

A script similar to `run_migrations_in_container.py` is provided, which automates copying this metadata file into this repository, if it's not already accessible to `manage.py` in the Django container. This is mentioned in the quick start notes.

There are _many_ things which we'll need/want to improve at some point. I will enter GitHub issues for the following items ASAP:
* Better generation of accession numbers and UUIDs from the HCA metadata, and/or matching what the data ingestion process will produce
* Building a more extensive provenance graph than simply using two levels from an HCA spreadsheet (donor and sample)
* Better indexing in Elasticsearch
* Functionality to drop the MongoDB and Elasticsearch databases from the Neo4j database (currently, the data import process is _not idempotent_, and duplicate data causes the import to fail)
* Communicating with Neo4j, MongoDB, and Elasticsearch in the Django application, and using those search results to drive the React frontend

mruffalo committed:
Add utils.normalize_whitespace function

mruffalo committed:
settings.py: add default configuration for backend services

These match the configuration provided in the development Docker config,
and can be overridden in production if/when required.

mruffalo committed:
Production Docker config: add new services

mruffalo committed:
Add data directory, ignore everything inside

mruffalo committed:
Docker config: disable Neo4j authentication for now

This is a bad idea in general, but the Neo4j container is only accessible
inside the private network set up by `docker-compose`. This is acceptable
for development and with our current "production" Docker configuration.

mruffalo committed:
Add metadata_import management command

mruffalo committed:
Fix lint warnings

mruffalo committed:
Move common Docker functionality to docker_util package

mruffalo committed:
Add load_hca_metadata.py

(which uses the new shared functionality in docker_util)

mruffalo committed:
README.md: add HCA metadata import to quick start section

mruffalo committed:
docker_util: add shared MANAGE_PY_COMMAND_TEMPLATE

mccalluc commented on data/.gitignore:
```diff
@@ -0,0 +1,2 @@
+*
+!.gitignore
```
Normally, I think it'd be simpler to ignore this directory from the top-level... but this way you can be assured that the directory exists, and that's something you want, right?

mruffalo replied:
Correct -- it would also suffice to create the directory in `load_hca_metadata.py`, but I kind of like the idea of "this directory is a standardized part of our application, but you should never commit the contents". I'm quite happy ignoring the directory entirely though.

mccalluc commented on docker/dev-common/docker-compose.yml:
```diff
@@ -32,6 +32,8 @@ services:
       - "7474:7474"
       - "7687:7687"
+    environment:
+      - NEO4J_AUTH=none
```
Right now, these two only differ at
```diff
13c13
<     image: "hubmap/data-portal-python-dev:latest"
---
>     build: django_app
```

As a tech-debt issue (not this PR), do you see any reason not to try cleaning this up with [multiple compose files](https://docs.docker.com/compose/extends/#multiple-compose-files)?

mruffalo replied:
Completely agree; thanks for the suggestion. I'll probably do that quite soon, which would allow removing the `dev-local-example` directory entirely.

mccalluc commented on docker_util/__init__.py:
```diff
@@ -0,0 +1,84 @@
+import json
+from pathlib import Path
+from subprocess import PIPE, run
+from typing import List
```
Have you considered using the [Docker Python SDK](https://docker-py.readthedocs.io/en/stable/)? Maybe for tech-debt rather than this PR.

mruffalo replied:
I did consider that, and specifically decided not to. All other things equal, I'd prefer to minimize the dependencies on the host system: right now, you only need Docker, a recent enough `docker-compose` binary, new enough Python, and no Python packages beyond the standard library. I could be convinced otherwise, though.

mccalluc replied:
That's reasonable. Particularly for a commandline tool I think this is fine. (If it were inside the web application I would have security and speed concerns.)

mccalluc commented on docker_util/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    str(REPOSITORY_PATH / 'hubmap/manage.py'),
+]
+
+def get_running_containers() -> List[str]:
```
Nifty: I haven't used this syntax before: Does one of your libraries reference this type annotation?

mruffalo replied:
I'm not 100% sure what you mean by "one of my libraries", but this is all Python standard library and syntax. It's not enforced at runtime with standard Python interpreters; it's just for improved code analysis (e.g. _telling_ PyCharm the return type of a function instead of making the tool infer it).

mccalluc commented on hubmap/dataportal/management/commands/_metadata_import.py:
```diff
@@ -0,0 +1,223 @@
+#!/usr/bin/env python3
```
The names differing in just a leading underscore could get confusing... Perhaps metadata_import_utils?

mruffalo replied:
That sounds reasonable, but the leading underscore tells Django "despite being in this directory, this Python file does not expose a `manage.py` command". It would be reasonable to have a single file instead of two, but I kind of liked the idea of separating "big ball of import code" and "Django entry point".

edit: so `_metadata_import_utils.py` would be great.

mccalluc commented on hubmap/dataportal/management/commands/_metadata_import.py:
```diff
@@ -0,0 +1,223 @@
+    # List first, then remove duplicates, then convert to nested structure
+    all_data = []
+    for i, row in donor_data.iterrows():
+        row_data = tuple(zip(row.index, value_adjust(row)))
```
`i` is currently unused?

mruffalo replied:
Right; `pandas.DataFrame.iterrows()` gives the row index and the row. I'll be happy to rename `i` to `_`.

mccalluc commented on hubmap/dataportal/management/commands/_metadata_import.py:
```diff
@@ -0,0 +1,223 @@
+        for key_str, value in row_data:
+            # TODO: probably skip if value is None
+            keys = key_str.split('.')
+            subdict = nested
```
Unnecessary? Just use `nested`?

mruffalo replied:
Hmm, yes. We can append to `nested_list` before repeatedly replacing the reference to `nested`.

mccalluc commented on hubmap/dataportal/management/commands/_metadata_import.py:
```diff
@@ -0,0 +1,223 @@
+            for key in keys[:-1]:
+                subdict = subdict[key]
+            subdict[keys[-1]] = value
+        nested_list.append(nested)
```
I'm having some trouble understanding this: An example in a doctest could be helpful.

mruffalo replied:
Will do.

mccalluc commented on load_hca_metadata.py:
```diff
@@ -0,0 +1,66 @@
+    return str(absolute_path_for_container)
+
+if __name__ == '__main__':
+    p = ArgumentParser()
```
Add `description`?

mruffalo replied:
Will do.

mccalluc reviewed (approved):
Looks good. Flagged some things (Avoiding copy and paste in config / using Docker SDK) that I think would be very good, but they don't need to block this PR. Some testing of the transformations of the input would be good... but how much effort goes into that depends on whether this should be considered a stopgap, or something we'll build upon?

Also:
- I'm a little worried about enriching the structures we have in Neo4J, and at the same time continuing to look at it as the definitive record of our metadata (ie, the thing we backup and rebuild from in case of catastrophic failure): I think the store of record should be as close as possible to the source data as possible. If the Neo4J will have relationships created by our code, then the Neo4J export will conflate the source data with our derived relationships.
- I'm taking this as a script that helps us get some data into the system for now... but not necessarily something that will be built on? In my mind, the input for this system will be something like the output from the `hubmap-sample-metadata` that I've been working on: Validated HCA JSON + RDF, rather than having this repo responsible for creating the graph from the raw CSV?

Let me know if you have the same eventual goal in mind, or if you are looking in a different direction...

mruffalo committed:
_metadata_import.py: add some docstrings

mruffalo committed:
Generalize _metadata_import.add_edge function

Don't hardcode 'donor_organism' and 'specimen_from_organism' node types.
I initially hardcoded these before annotating the node types, and forgot
to fix this earlier.

mruffalo committed:
Make dict unflattening into a function with a docstring

mruffalo committed:
Rename _metadata_import.py -> _metadata_import_utils.py

(as per Chuck's suggestion)

mruffalo committed:
load_hca_metadata.py: add description to ArgumentParser

mruffalo committed:
_metadata_import_utils.py: ignore "unused" import of pprint

Apparently flake8 isn't smart enough to understand that it's used in the
doctest of 'dotted_keys_to_nested_dict'.

mccalluc commented on README.md:
```diff
@@ -27,8 +27,8 @@ docker-compose up
-run_migrations_in_container.py
-# TODO: Load sample data into DB
+./run_migrations_in_container.py
+./load_hca_metadata.py path_to_metadata_from_HCA.tsv
```
Can you add a `wget` line to explain where precisely this metadata TSV comes from?

mruffalo replied:
I'm not sure -- the metadata download links on HCA's portal have some scripting functionality to prepare a download (or something like that). I don't know how stable the URLs are, so I'll add a note to download metadata with a web browser.

mccalluc commented on docker_util/__init__.py:
```diff
@@ -0,0 +1,84 @@
+def run_in_django_container(command: List[str], pretend: bool):
+    containers = get_running_containers()
+    if not containers:
+        label_str = '\n'.join(f'\t{label}'for label in CONTAINER_LABELS)
```
```suggestion
        label_str = '\n'.join(f'\t{label}' for label in CONTAINER_LABELS)
```

mccalluc commented on docker_util/__init__.py:
```diff
@@ -0,0 +1,84 @@
+            return proc.stdout.strip().decode('utf-8')
+
+def run_in_django_container(command: List[str], pretend: bool):
+    containers = get_running_containers()
```
Maybe rename to make it more explicit? Or this might be clear enough from context.
```suggestion
    django_containers = get_running_django_containers()
```

mccalluc commented on docker_util/__init__.py:
```diff
@@ -0,0 +1,84 @@
+        raise EnvironmentError(message)
+
+    if len(containers) > 1:
+        print('Warning: multiple containers found ({}). Using first.'.format(', '.join(containers)))
```
Would it be safer to fail here?

mccalluc commented on hubmap/dataportal/management/commands/_metadata_import_utils.py:
```diff
@@ -0,0 +1,259 @@
+        'a.f': 'another', \
+        'g.h': None, \
+    }
+    >>> pprint(dotted_keys_to_nested_dict(d))
```
Not sure why pprint is needed?

mruffalo replied:
I originally wanted a multi-line display of the dict, but `pprint` uses a single line if the output is within a certain width. I'll remove that.

mccalluc reviewed (approved):
Thanks for the tweaks. Since it's in the same space, are you on board with plan described in #132? Comment there if that doesn't seem right.

mruffalo committed:
_metadata_import_utils.py: drop usage of pprint

The `repr` of this dict matches the output of `pprint` since it's so small.

mruffalo committed:
README.md: add note to download metadata from HCA

mruffalo committed:
docker_util: tweaks from Chuck's comments
